gnucash/gnucash-commands.cpp

```diff
@@ -434,3 +546,141 @@ Gnucash::report_list (void)
     scm_boot_guile (0, nullptr, scm_report_list, NULL);
     return 0;
 }
+
+// scripting code follows:
+
+static void
+run_guile_cli (void *data, [[maybe_unused]] int argc, [[maybe_unused]] char **argv)
+{
+    auto args = static_cast<scripting_args*>(data);
+    if (args->script)
+    {
+        PINFO ("Running script from %s... ", args->script->c_str());
+        scm_c_primitive_load (args->script->c_str());
+    }
+    if (args->interactive)
+    {
+        std::cout << _("Welcome to Gnucash Interactive Guile Session") << std::endl;
+        std::vector<const char*> modules =
+            { "gnucash core-utils", "gnucash engine", "gnucash app-utils", "gnucash report",
+              "system repl repl", "ice-9 readline" };
+        auto show_and_load = [](const auto& mod)
+        {
+            std::cout << bl::format ("(use-modules ({1}))") % mod << std::endl;
+            scm_c_use_module (mod);
+        };
+        std::for_each (modules.begin(), modules.end(), show_and_load);
+        scm_c_eval_string ("(activate-readline)");
+        scm_c_eval_string ("(start-repl)");
+    }
+    cleanup_and_exit_with_save ();
+}
+
+
+#ifdef HAVE_PYTHON_H
+static void
+python_cleanup (PyConfig& config, PyStatus& status)
+{
+    if (qof_book_session_not_saved (gnc_get_current_book()))
+        std::cerr << _("Book is readonly. Unsaved changes will be lost.") << std::endl;
+    gnc_clear_current_session ();
+
+    PyConfig_Clear(&config);
+    if (PyStatus_IsExit(status))
+        gnc_shutdown_cli (status.exitcode);
+
+    Py_ExitStatusException(status);
```

Comment:
> [`Py_ExitStatusException` ](https://docs.python.org/3/c-api/init_config.html#c.PyStatus_Exception) calls `exit` either way, so ISTM you want to use `gnc_shutdown_cli` to ensure that the session gets closed. We don't want to leave a lockfile just because the script barfed or the user passed bad options to python.

In this particular case the session is already cleared via `gnc_clear_current_session ()` a couple lines up, so, it'll simply call engine shutdown procedures instead.